### PR TITLE
[Merged by Bors] - feat(RingTheory/StandardSmooth): composition and localization

### PR DESCRIPTION
**Pull request by chrisflav**
Constructs submersive presentations for composition and localization away from one element.

This contribution was created as part of the AIM workshop "Formalizing algebraic geometry"
in June 2024.

---

- [x] depends on: #16727

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/RingTheory/Smooth/StandardSmooth.lean [hunk 1]
```diff
@@ -190,7 +186,159 @@ lemma ofBijectiveAlgebraMap_jacobian (h : Function.Bijective (algebraMap R S)) :
...
+  map_inj _ _ h := h
+  relations_finite := inferInstanceAs <| Finite Unit
+
+instance : Fintype (localizationAway S r).rels :=
+  inferInstanceAs (Fintype Unit)
+
+instance : DecidableEq (localizationAway S r).rels :=
+  inferInstanceAs (DecidableEq Unit)
+
+lemma localizationAway_jacobiMatrix :
```
Comment (jcommelin) [inline Mathlib/RingTheory/Smooth/StandardSmooth.lean:209]: Should this be a simp lemma?
```suggestion
@[simp]
lemma localizationAway_jacobiMatrix :
```

File: Mathlib/RingTheory/Smooth/StandardSmooth.lean [hunk 2]
```diff
@@ -190,7 +186,159 @@ lemma ofBijectiveAlgebraMap_jacobian (h : Function.Bijective (algebraMap R S)) :
...
+    (localizationAway S r).jacobiMatrix = Matrix.diagonal (fun () ↦ MvPolynomial.C r) := by
+  have h : (pderiv ()) (C r * X () - 1) = C r := by simp
+  ext (i : Unit) (j : Unit) : 1
+  rwa [jacobiMatrix_apply]
+
+lemma localizationAway_jacobian : (localizationAway S r).jacobian = algebraMap R S r := by
```
Comment (jcommelin) [inline Mathlib/RingTheory/Smooth/StandardSmooth.lean:215]: And this one?
```suggestion
@[simp]
lemma localizationAway_jacobian : (localizationAway S r).jacobian = algebraMap R S r := by
```

File: Mathlib/RingTheory/Smooth/StandardSmooth.lean [hunk 3]
```diff
@@ -190,7 +186,159 @@ lemma ofBijectiveAlgebraMap_jacobian (h : Function.Bijective (algebraMap R S)) :
...
+    erw [MvPolynomial.eval₂_C]
+  · simp [hp, hq]
+  · simp only [map_mul, rename_X, eval₂_mul, hp, eval₂_X]
+    erw [Generators.comp_val]
+    simp
+
+end
+
+/-- The jacobian of the composition of presentations is the product of the jacobians. -/
+lemma comp_jacobian_eq_jacobian_smul_jacobian : (Q.comp P).jacobian = P.jacobian • Q.jacobian := by
```
Comment (jcommelin) [inline Mathlib/RingTheory/Smooth/StandardSmooth.lean:326]: And this one?
```suggestion
@[simp]
lemma comp_jacobian_eq_jacobian_smul_jacobian : (Q.comp P).jacobian = P.jacobian • Q.jacobian := by
```

**Comment by chrisflav** [comment]
I made the suggested ones (and some existing `jacobian` lemmas) `simp`. I believe these will rarely trigger, but I don't think it hurts.

**Review by jcommelin — COMMENTED** [review]
maintainer merge

File: Mathlib/RingTheory/Smooth/StandardSmooth.lean [hunk 4]
```diff
@@ -190,7 +187,162 @@ lemma ofBijectiveAlgebraMap_jacobian (h : Function.Bijective (algebraMap R S)) :
...
+
+private lemma jacobiMatrix_comp_₁₁_det :
+    (aeval (Q.comp P).val) (Q.comp P).jacobiMatrix.toBlocks₁₁.det = Q.jacobian := by
+  rw [jacobian_eq_jacobiMatrix_det, AlgHom.map_det (aeval (Q.comp P).val), RingHom.map_det]
+  congr
+  ext i j : 1
+  simp only [Matrix.map_apply, RingHom.mapMatrix_apply, ← Q.jacobiMatrix_comp_inl_inl P]
+  apply aeval_sum_elim
+
+omit [Fintype Q.rels]
```
Comment (riccardobrasca) [inline Mathlib/RingTheory/Smooth/StandardSmooth.lean:297]: Can you create sections instead of using `omit`?

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

**Pull request closed without merging**